docs/src/userguide/extension_types.rst

```diff
@@ -530,6 +530,107 @@ statically sized freelist of ``N`` instances for a given type.  Example::
     penguin = None
     penguin = Penguin('fish 2')  # does not need to allocate memory!
 
+.. _existing-pointers-instantiation:
+
+Instantiation from existing C/C++ pointers
+===========================================
+
+It is quite common to want to instantiate an extension class from an existing
+(pointer to a) data structure, often as returned by external C/C++ functions.
+
+As extension classes can only accept Python objects as arguments in their
+contructors, this necessitates the use of factory functions. For example, ::
+
+    from libc.stdlib cimport malloc, free
+
+    # Example C struct
+    ctypedef struct my_c_struct:
+        int a
+        int b
+
+
+    cdef class WrapperClass:
+        """A wrapper class for a C/C++ data structure"""
+        cdef my_c_struct *_ptr
+        cdef bint ptr_owner
+
+        def __cinit__(self, owner=False):
```

Comment:
I wouldn't recommend this approach, for two reasons.
1) The argument is Python user visible and controllable, which makes it less safe from a C point of view.
2) Passing a keyword argument through Python (or, in fact, any arguments at all) is not the most efficient thing you can do, and wrapper instantiation is often time critical.

Apart from passing pointers, one of the main reasons why people use factory functions is performance, because calling `__new__()` and setting attributes directly can be much faster than going through the complete Python object instantiation and especially passing arguments through the constructor, but you wouldn't want to spread that ugly instantiation code all through your code base. Thus, factory functions/methods.

I thus consider it good practice to
1) Pass no arguments into `__cinit__()` (if possible) and only bring the object into the minimally sane state there. You also don't need to set any pointer attributes to `NULL` there, since the object memory is always zeroed out after allocation.
2) If instantiation from Python is allowed, do the non-trivial state initialisation (and thorough input checking, etc.) from `__init__()`, also to nicely support subclassing. In your example, you could accept a tuple with two integers as arguments to show that, and create a new struct from their values.
3) Create cdef factory methods for time citical or non-Python input initialisation, and let them do the complete state setup with all attribute setting. Since they can only be called from C(ython) code, they can focus on speed rather than safety and don't need to do as much input checking.

Reply:
All good points, thank you. Have updated without any arguments to `__cinit__`.